ribbon/sign.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#----------------------------------------------------------------------------
+# Created By: Steven (steven@ribbon.finance)
+# Created Date: 04/04/2022
+# version ='0.01'
+# ---------------------------------------------------------------------------
+''' Module to generate signature to bid '''
+# ---------------------------------------------------------------------------
+
+# ---------------------------------------------------------------------------
+# Imports
+# ---------------------------------------------------------------------------
+import eth_keys
+from encode import TypedDataEncoder
+
+# ---------------------------------------------------------------------------
+# Signature Generator
+# ---------------------------------------------------------------------------
+class SignatureGenerator:
+  '''
+  Object to generate bid signature
+
+  Args:
+      privateKey (str): Private key of the user in hex format with 0x prefix
+
+  Attributes:
+      signer (object): Instance of signer to generate signature
+  '''
+  def __init__(self, privateKey: str) -> None:
+    self.signer = eth_keys.keys.PrivateKey(bytes.fromhex(privateKey[2:]))
+
+  def signMessage(self, messageHash: str) -> dict:
+    '''Sign a hash message using the signer object
+
+    Args:
+        messageHash (str): Message to signed in hex format with 0x prefix
+
+    Returns:
+        signature (dict): Signature split into v, r, s components
+    '''
+    signature = self.signer.sign_msg_hash(bytes.fromhex(messageHash[2:]))
+
+    return {
+      'v': signature.v + 27,
+      'r': hex(signature.r),
+      's': hex(signature.s)
+    }
+
+  def _signTypedDataV4(self, domain: dict, types: dict, value: dict) -> str:
+    '''Sign a hash of typed data V4 which follows EIP712 convention:
+    https://eips.ethereum.org/EIPS/eip-712
+    
+    Args:
+        domain (dict): Dictionary containing domain parameters including
```

Comment:
Having a generic dictionary accepted here would require people to understand how they look like from somewhere else, and can create issues.

We may do this directly in the code by having something like
```
from dataclasses import dataclass, asdict

@dataclass
class DomainParameters:
    """Domain Paylod for signatures"""
    name: str
    chainId: int
    verifyingContract: str
    version: int = 1
```

And then we can have people do:
```python
domain = DomainParameters(name='RIBBON SWAP', chainId=42, verifyingContract='0x58848824baEb9678847aF487CB02EAba782FECB5')
asdict(domain)

# Output
{'name': 'RIBBON SWAP',
 'chainId': 42,
 'verifyingContract': '0x58848824baEb9678847aF487CB02EAba782FECB5',
 'version': 1}
```

The same can be done for the values, which maybe we can just call `Payload` or `BidPayload` or just `Bid`

Reply:
Or we can do 
```
def _signTypedDataV4(self, domain: DomainParameters, value: Bid, types: dict = BID_TYPES) -> str
```
and internally to the function we translate those to dicts

Reply:
Added dataclasses for `Domain`, `Bid` and `SignedBid`